qmp/socket.go

```diff
@@ -157,7 +170,14 @@ func (mon *SocketMonitor) Connect() error {
 // Events should only be called once per Socket.  If used with a qemu.Domain,
 // qemu.Domain.Events should be called to retrieve events instead.
 func (mon *SocketMonitor) Events(context.Context) (<-chan Event, error) {
-	atomic.AddInt32(mon.listeners, 1)
+	mon.lmu.Lock()
+	defer mon.lmu.Unlock()
+
+	select {
+	case <-mon.noListeners:
+		mon.noListeners = make(chan struct{})
```

Comment:
Interesting technique. But I don't think it's at all clear to the reader, and the presence of that mutex makes it hard to reason about.

I think a better approach would be when the last listener disconnects, attach a goroutine to the channel with a simple read loop, to consume all further messages on the channel. That way the sender can send the error into the void and close the channel, at which point the reader goroutine can also exit.

Reply:
That is a zillion times simpler :sweat_smile: